mava/systems/ff_ippo_rware.py

```diff
@@ -312,44 +390,70 @@ def learner_fn(learner_state: LearnerState) -> ExperimentOutput:
     """Initialise learner_fn, network, optimiser, environment and states."""
     # Get available TPU cores.
     n_devices = len(jax.devices())
+
     # Get number of actions and agents.
     num_actions = int(env.action_spec().num_values[0])
     num_agents = env.action_spec().shape[0]
     config["num_agents"] = num_agents
+
     # PRNG keys.
     rng, rng_p = rngs
 
     # Define network and optimiser.
-    network = ActorCritic(num_actions)
-    optim = optax.chain(
+    actor_network = Actor(num_actions)
+    critic_network = Critic()
+    actor_optim = optax.chain(
+        optax.clip_by_global_norm(config["max_grad_norm"]),
+        optax.adam(config["actor_lr"], eps=1e-5),
+    )
+    critic_optim = optax.chain(
         optax.clip_by_global_norm(config["max_grad_norm"]),
-        optax.adam(config["lr"], eps=1e-5),
+        optax.adam(config["critic_lr"], eps=1e-5),
     )
 
     # Initialise observation: Select only obs for a single agent.
     init_x = env.observation_spec().generate_value()
     init_x = jax.tree_util.tree_map(lambda x: x[0], init_x)
     init_x = jax.tree_util.tree_map(lambda x: x[None, ...], init_x)
 
-    # initialise params and optimiser state.
-    params = network.init(rng_p, init_x)
-    opt_state = optim.init(params)
+    # Initialise actor params and optimiser state.
+    actor_params = actor_network.init(rng_p, init_x)
+    actor_opt_state = actor_optim.init(actor_params)
+
+    # Initialise critic params and optimiser state.
+    critic_params = critic_network.init(rng_p, init_x)
+    critic_opt_state = critic_optim.init(critic_params)
 
     # Vmap network apply function over number of agents.
-    vmapped_network_apply_fn = jax.vmap(network.apply, in_axes=(None, 1), out_axes=(1, 1))
+    vmapped_actor_network_apply_fn = jax.vmap(
+        actor_network.apply,
+        in_axes=(None, 1),
+        out_axes=(1),
+    )
+    vmapped_critic_network_apply_fn = jax.vmap(
+        critic_network.apply,
+        in_axes=(None, 1),
+        out_axes=(1),
+    )
+
+    # Pack apply and update functions.
+    apply_fns = (vmapped_actor_network_apply_fn, vmapped_critic_network_apply_fn)
+    update_fns = (actor_optim.update, critic_optim.update)
 
     # Get batched iterated update and replicate it to pmap it over cores.
-    learn = get_learner_fn(env, vmapped_network_apply_fn, optim.update, config)
+    learn = get_learner_fn(env, apply_fns, update_fns, config)
     learn = jax.pmap(learn, axis_name="device")
 
     # Broadcast params and optimiser state to cores and batch.
     broadcast = lambda x: jnp.broadcast_to(x, (n_devices, config["update_batch_size"]) + x.shape)
-    params = jax.tree_map(broadcast, params)
-    opt_state = jax.tree_map(broadcast, opt_state)
+    actor_params = jax.tree_map(broadcast, actor_params)
+    actor_opt_state = jax.tree_map(broadcast, actor_opt_state)
+    critic_params = jax.tree_map(broadcast, critic_params)
+    critic_opt_state = jax.tree_map(broadcast, critic_opt_state)
```

Comment:
I _think_ this should work and be a bit cleaner, since `Params`/`OptStates` is just a pytree
```suggestion
    params = Params(actor_params, critic_params)
    opt_states = OptStates(actor_opt_state, critic_opt_state)
    
    params = jax.tree_map(broadcast, params)
    opt_states = jax.tree_map(broadcast, opt_states)
```